File: api/backtest.py
"""
Vercel Serverless Function for FluxBack Backtesting API
Note: This requires the Python bindings to be built and available
"""

import json
import sys
import os
from typing import Dict, Any

def handler(request):
    """
    Handle backtest requests
    For Vercel, this will be called via serverless function
    """
    try:
        # Parse request body
        if hasattr(request, 'json'):
            data = request.json
        else:
            data = json.loads(request.body) if hasattr(request, 'body') else {}
        
        strategy_yaml = data.get('strategy', '')
        csv_data = data.get('data', '')
        
        if not strategy_yaml or not csv_data:
            return {
                'statusCode': 400,
                'body': json.dumps({'error': 'Missing strategy or data'})
            }
        
        # Try to import fluxback_py (Python bindings)
        try:
            import fluxback_py
        except ImportError:
            # Fallback: return instructions for local setup
            return {
                'statusCode': 503,
                'body': json.dumps({
                    'error': 'Python bindings not available on server',
                    'message': 'Please run backtests locally or set up Python bindings',
                    'instructions': 'Clone the repo and run: python -m pip install -e .'
                })
            }
        
        # Save temporary files
        import tempfile
        with tempfile.NamedTemporaryFile(mode='w', suffix='.yaml', delete=False) as f:
            f.write(strategy_yaml)
            strategy_path = f.name
        
        with tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False) as f:
            f.write(csv_data)
            data_path = f.name
        
        # Run backtest
        result = fluxback_py.run_backtest(strategy_path, data_path)
        
        # Cleanup
        os.unlink(strategy_path)
        os.unlink(data_path)
        
        return {
            'statusCode': 200,
            'body': json.dumps(result)
        }
        
    except Exception as e:
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }

# Vercel serverless function entry point
def main(request):
    return handler(request)

File: api/backtest.js
// Vercel Serverless Function - JavaScript version
// This is a fallback that provides instructions since C++ can't run on Vercel

export default async function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const { strategy, data } = req.body;

  if (!strategy || !data) {
    return res.status(400).json({ error: 'Missing strategy or data' });
  }

  // Since we can't run C++ on Vercel serverless, return instructions
  // For a real implementation, you'd need:
  // 1. WebAssembly compilation of the C++ code, OR
  // 2. A separate server with Python bindings, OR
  // 3. Client-side execution via WebAssembly
  
  return res.status(503).json({
    error: 'Online backtesting requires WebAssembly setup',
    message: 'The C++ engine needs to be compiled to WebAssembly for browser execution',
    instructions: {
      local: 'Clone the repo and run locally: git clone https://github.com/NikhilGolla72/GollaNikhil-PaceStock.git && cd GollaNikhil-PaceStock && ./run_demo.ps1',
      webassembly: 'To enable online backtesting, compile the C++ code to WebAssembly using Emscripten'
    },
    // Return sample structure for UI testing
    sample_result: {
      total_return_pct: -0.02,
      sharpe_ratio: -6.24,
      win_rate_pct: 23.53,
      total_trades: 34,
      winning_trades: 8,
      losing_trades: 26,
      max_drawdown_pct: 0.02,
      profit_factor: 0.22,
      initial_cash: 100000,
      final_cash: 99980.72
    }
  });
}

